### PR TITLE
[ONNX] GlobalLpPool operator

### PR DESCRIPTION
**Pull request by arogowie-intel**
Involves slight changes to internal reduction operations API to prevent code duplication for calculating L1,L2 norms.

Note that norms are [entrywise](https://en.wikipedia.org/wiki/Matrix_norm#%22Entrywise%22_matrix_norms), that is the input tensors's axes to be reduced are _flattened_ and treated like long vector.

### CODE REVIEW

File: src/ngraph/frontend/onnx_import/utils/reduction.hpp [hunk 1]
```diff
@@ -57,22 +57,24 @@ namespace ngraph
...
+            /// \tparam     ReductionFunction   Functor defining arithmetic reduction operation
+            ///                                 (e.g. Min, Max, SUm, Product).
             ///
             /// \return     nGraph node equivalent of the ONNX operation.
             ///
-            template <class OnnxOperator,
-                      typename std::enable_if<std::is_base_of<ngraph::op::util::ArithmeticReduction,
-                                                              OnnxOperator>::value,
-                                              int>::type = 0>
+            template <class ReductionFunction>
```
Comment (postrational) [inline src/ngraph/frontend/onnx_import/utils/reduction.hpp]: Why are we removing `std::enable_if` here?

File: src/ngraph/frontend/onnx_import/utils/reduction.hpp [hunk 2]
```diff
@@ -57,22 +57,24 @@ namespace ngraph
...
+            /// \brief      Create an nGraph version of an ONNX reduction
+            ///             operation.
             ///
-            /// \tparam     OnnxOperator  Class of an nGraph ArithmeticReduction operation
-            ///                           (e.g. Min, Max, SUm, Product).
+            /// \param[in]  node                The node representing incoming ONNX operation.
+            /// \param[in]  ng_input            The input (nGraph) Tensor.
+            /// \param[in]  reduction_function  The reduction function.
+            ///
+            /// \tparam     ReductionFunction   Functor defining arithmetic reduction operation
```
Comment (postrational) [inline src/ngraph/frontend/onnx_import/utils/reduction.hpp]: What is the functor signature?

File: src/ngraph/frontend/onnx_import/utils/reduction.hpp [hunk 3]
```diff
@@ -57,22 +57,24 @@ namespace ngraph
...
+            ///                                 (e.g. Min, Max, SUm, Product).
```
Comment (postrational) [inline src/ngraph/frontend/onnx_import/utils/reduction.hpp]: Typo in `SUm`

File: src/ngraph/frontend/onnx_import/op/reduce.hpp [hunk 4]
```diff
@@ -192,8 +222,13 @@ namespace ngraph
...
                 inline NodeVector reduce_sum(const Node& node)
                 {
-                    return {reduction::make_ng_reduction_op<ngraph::op::Sum>(
-                        node, node.get_ng_inputs().at(0))};
+                    return {reduction::make_ng_reduction_op(
+                        node,
+                        node.get_ng_inputs().at(0),
+                        [](const std::shared_ptr<ngraph::Node>& input,
+                           const ngraph::AxisSet& reduction_axes) {
+                            return std::make_shared<ngraph::op::Sum>(input, reduction_axes);
```
Comment (postrational) [inline src/ngraph/frontend/onnx_import/op/reduce.hpp]: Could you just use `std::make_shared<ngraph::op::Sum>` in place of the lambda?

File: src/ngraph/frontend/onnx_import/op/reduce.hpp [hunk 5]
```diff
@@ -158,8 +178,13 @@ namespace ngraph
...
                 inline NodeVector reduce_min(const Node& node)
                 {
-                    return {reduction::make_ng_reduction_op<ngraph::op::Min>(
-                        node, node.get_ng_inputs().at(0))};
+                    return {reduction::make_ng_reduction_op(
+                        node,
+                        node.get_ng_inputs().at(0),
+                        [](const std::shared_ptr<ngraph::Node>& input,
+                           const ngraph::AxisSet& reduction_axes) {
+                            return std::make_shared<ngraph::op::Min>(input, reduction_axes);
```
Comment (postrational) [inline src/ngraph/frontend/onnx_import/op/reduce.hpp]: Could you just `std::make_shared<ngraph::op::Min>` in place of the lambda?

File: src/ngraph/frontend/onnx_import/op/reduce.hpp [hunk 6]
```diff
@@ -175,8 +200,13 @@ namespace ngraph
...
                 inline NodeVector reduce_prod(const Node& node)
                 {
-                    return {reduction::make_ng_reduction_op<ngraph::op::Product>(
-                        node, node.get_ng_inputs().at(0))};
+                    return {reduction::make_ng_reduction_op(
+                        node,
+                        node.get_ng_inputs().at(0),
+                        [](const std::shared_ptr<ngraph::Node>& input,
+                           const ngraph::AxisSet& reduction_axes) {
+                            return std::make_shared<ngraph::op::Product>(input, reduction_axes);
```
Comment (postrational) [inline src/ngraph/frontend/onnx_import/op/reduce.hpp]: Could you just `std::make_shared<ngraph::op::Product>` in place of the lambda?

File: src/ngraph/frontend/onnx_import/op/reduce.hpp [hunk 7]
```diff
@@ -128,8 +143,13 @@ namespace ngraph
...
                 inline NodeVector reduce_max(const Node& node)
                 {
-                    return {reduction::make_ng_reduction_op<ngraph::op::Max>(
-                        node, node.get_ng_inputs().at(0))};
+                    return {reduction::make_ng_reduction_op(
+                        node,
+                        node.get_ng_inputs().at(0),
+                        [](const std::shared_ptr<ngraph::Node>& input,
+                           const ngraph::AxisSet& reduction_axes) {
+                            return std::make_shared<ngraph::op::Max>(input, reduction_axes);
```
Comment (postrational) [inline src/ngraph/frontend/onnx_import/op/reduce.hpp]: Could you just `std::make_shared<ngraph::op::Max>` in place of the lambda?

File: src/ngraph/frontend/onnx_import/op/reduce.hpp [hunk 8]
```diff
@@ -90,8 +105,10 @@ namespace ngraph
                 ///
                 inline NodeVector reduce_l1(const Node& node)
                 {
-                    auto abs_node = std::make_shared<ngraph::op::Abs>(node.get_ng_inputs().at(0));
-                    return {reduction::make_ng_reduction_op<ngraph::op::Sum>(node, abs_node)};
+                    return {reduction::make_ng_reduction_op(
+                        node,
+                        node.get_ng_inputs().at(0),
+                        std::bind(norm::l1_norm, std::placeholders::_1, std::placeholders::_2))};
```
Comment (postrational) [inline src/ngraph/frontend/onnx_import/op/reduce.hpp]: Can we avoid `bind` here?

File: src/ngraph/frontend/onnx_import/op/reduce.hpp [hunk 9]
```diff
@@ -107,12 +124,10 @@ namespace ngraph
...
-                    NodeVector ng_inputs{node.get_ng_inputs()};
-                    auto square_node =
-                        std::make_shared<ngraph::op::Multiply>(ng_inputs.at(0), ng_inputs.at(0));
-                    auto sum_node =
-                        reduction::make_ng_reduction_op<ngraph::op::Sum>(node, square_node);
-                    return {std::make_shared<ngraph::op::Sqrt>(sum_node)};
+                    return {reduction::make_ng_reduction_op(
+                        node,
+                        node.get_ng_inputs().at(0),
+                        std::bind(norm::l2_norm, std::placeholders::_1, std::placeholders::_2))};
```
Comment (postrational) [inline src/ngraph/frontend/onnx_import/op/reduce.hpp]: Can we avoid `bind` here?

Comment (arogowie-intel) [reply src/ngraph/frontend/onnx_import/op/reduce.hpp]: Yup.

File: src/ngraph/frontend/onnx_import/op/reduce.hpp (hunk 8 above)
Comment (arogowie-intel) [reply src/ngraph/frontend/onnx_import/op/reduce.hpp]: Yup.

File: src/ngraph/frontend/onnx_import/op/reduce.hpp (hunk 7 above)
Comment (arogowie-intel) [reply src/ngraph/frontend/onnx_import/op/reduce.hpp]: Yup.

File: src/ngraph/frontend/onnx_import/op/reduce.hpp (hunk 6 above)
Comment (arogowie-intel) [reply src/ngraph/frontend/onnx_import/op/reduce.hpp]: Yup.

File: src/ngraph/frontend/onnx_import/op/reduce.hpp (hunk 5 above)
Comment (arogowie-intel) [reply src/ngraph/frontend/onnx_import/op/reduce.hpp]: Yup.

File: src/ngraph/frontend/onnx_import/op/reduce.hpp (hunk 4 above)
Comment (arogowie-intel) [reply src/ngraph/frontend/onnx_import/op/reduce.hpp]: Yup.

File: src/ngraph/frontend/onnx_import/utils/reduction.hpp (hunk 2 above)
Comment (arogowie-intel) [reply src/ngraph/frontend/onnx_import/utils/reduction.hpp]: There is now alias `ReductionFunction` which defines function object parameter type.

File: src/ngraph/frontend/onnx_import/utils/reduction.hpp (hunk 1 above)
Comment (arogowie-intel) [reply src/ngraph/frontend/onnx_import/utils/reduction.hpp]: We're no longer force type safety by using object from specific type hierarchy.

**Pull request merged**